### PR TITLE
[Bug Fix] tril/triu: masked-out inf/NaN becomes NaN instead of 0 (#52038)

### LINKED ISSUE #52038: [ttnn]: tril/triu are composed as a mask multiply, so an inf in the masked region becomes NaN instead of 0 (fp32)
### Component / Area

ttnn

### Issue Type (optional)

Bad Outputs

### Observed

Measured on ttsim, Tenstorrent's own simulator, in **float32**. I do not have Tenstorrent silicon; a maintainer pointed me at ttsim in #51845 and it runs the real kernel.

Scope, up front, because it is narrower than the composition alone would suggest: **this reproduces on the float32 path and not on the bfloat16 one.** On a 32x32 tensor filled with `-inf`, fp32 turns 496 of the 496 masked cells into NaN; bf16 turns 0 of 496 into NaN. I cannot tell whether the bf16 behaviour is the device's or the simulator's, so I am only claiming fp32.

`tril` selects the lower triangular region and **zeroes** the rest. It is composed as a multiply against a 0/1 mask (`unary_composite_op.cpp:308-318`):

```cpp
Tensor tril(const Tensor& input_a, std::int32_t diag, const std::optional<MemoryConfig>& output_mem_config) {
    Tensor index_l = ttnn::index_tril<::bfloat16>(
        input_a.logical_shape(), input_a.padded_shape(), diag,
        DataType::BFLOAT16, Layout::TILE, input_a.device(),
        output_mem_config.value_or(input_a.memory_config()));
    return ttnn::multiply(input_a, index_l, std::nullopt, output_mem_config);
}
```

Zeroing and multiplying by zero are not the same operation. In IEEE-754, `0 * ±inf` is `NaN`, and `0 * NaN` is `NaN`. So every masked-out element that happens to hold an infinity or a NaN comes out as **NaN** rather than `0`.

`triu` immediately below has the same shape with `index_triu`.

**Where this actually bites.** Causal attention masks are commonly built with `-inf` in the positions to be suppressed, and tril/triu are the natural way to build or apply them. Measured on a 32x32 float32 tensor filled with `-inf`:

| fill | cells the mask must zero | come out 0 | **come out NaN** |
|---|---|---|---|
| **-inf** | 496 | 0 | **496** |
| **+inf** | 496 | 0 | **496** |
| -1e30 (control) | 496 | **496** | 0 |
| -1.0 (control) | 496 | **496** | 0 |

**Every masked cell, not some of them.** The two controls are the point: with a large finite fill the mask works perfectly, so this is the infinity specifically and not the masking in general. The retained triangle is untouched in all four cases.

The rate is 100%, so the size does not matter — a 128x128 causal mask loses all 8,128 masked positions the same way. The NaNs then propagate through the softmax, so a whole row is lost rather than one element being slightly off.

`torch.tril` does not behave this way: it writes zeros into the masked region rather than scaling it, so on the same input the registered golden and the kernel disagree.

**On test coverage.** The sweep at `tests/sweep_framework/sweeps/eltwise/unary/tril/tril.py` generates inputs with `torch_random(low=-100, high=100)`, so it never produces an infinity or a NaN. That is why this has not surfaced: the failing region sits outside what the sweep samples, rather than the sweep accepting it.

### Expected

The masked region should be exactly `0` for every input, including `±inf` and `NaN`, matching `torch.tril` and `torch.triu`.

The invariant, checkable without any golden table:

```
tril(x)[i][j] == 0            for every j > i + diag
tril(x)[i][j] == x[i][j]      otherwise
```

A select is the natural composition and it already exists in ttnn: `ttnn::where(mask, input, 0.0f)` keeps the kept region bit-exact and writes a true zero into the rest, with no arithmetic on the masked values at all. That also removes the multiply's rounding on the kept side, which is a small bonus rather than the point.

If the multiply is preferred for performance reasons, the masked region still needs a guard so non-finite values do not survive it.

### 1. Steps (exact commands)

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)

N = 32
x = torch.full((N, N), float("-inf"), dtype=torch.float32)   # fp32: see scope note
t = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
out = ttnn.to_torch(ttnn.tril(t))

upper = [out[i][j].item() for i in range(N) for j in range(N) if j > i]
print("masked cells:", len(upper))                                  # 496
print("NaN         :", sum(1 for v in upper if v != v))             # 496  -- expected 0
print("zero        :", sum(1 for v in upper if v == 0.0))           # 0    -- expected 496
print("torch NaNs  :", torch.tril(x).isnan().sum().item())          # 0

# control: same shape, large finite fill -> the mask works
y = torch.full((N, N), -1e30, dtype=torch.float32)
ty = ttnn.from_torch(y, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
oy = ttnn.to_torch(ttnn.tril(ty))
print("control NaN :", sum(1 for i in range(N) for j in range(N)
                           if j > i and oy[i][j].item() != oy[i][j].item()))   # 0
```

Minimal single-element version:

```python
x = torch.tensor([[1.0, float("inf")], [1.0, 1.0]], dtype=torch.float32)
t = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
print(ttnn.to_torch(ttnn.tril(t)))   # [[1, nan], [1, 1]]  -- expected [[1, 0], [1, 1]]
print(torch.tril(x))                 # [[1, 0], [1, 1]]
```

Swapping `torch.float32`/`ttnn.float32` for bfloat16 in the snippet above gives 0 NaN and 496 zeros — the bf16 path does not reproduce, which is why the claim is scoped to fp32.

### 2. Input data / link or description

No data file. Inputs are constructed in the snippet: a 32x32 float32 tensor filled with `-inf`, plus a large-finite control. The 128x128 causal-mask case behaves identically, since the failure rate is 100%.

### 3. Frequency

Deterministic — 100% for any non-finite value landing in the masked region. It is a property of the composition, not a race or a flake.

### 1. Software Versions

Deterministic — 100% for any non-finite value landing in the masked region. It is a property of the composition, not a race or a flake.

### 2. Hardware Details

**No Tenstorrent silicon.** Everything above is measured on **ttsim** running the real kernel, in a stock `tt-metalium` release container, float32. The bfloat16 path does not reproduce there and I make no claim about it.

What that does and does not affect:

- The composition (`ttnn::multiply` against a 0/1 mask) is read from `unary_composite_op.cpp:308-318` in `main`. Not device-dependent.
- `0 * ±inf = NaN` and `0 * NaN = NaN` are IEEE-754. Not device-dependent.
- The 8,128 figure is a direct count of the masked positions in a 128x128 causal layout, not a simulation.
- The question of whether the SFPU multiply follows IEEE or flushes is already answered on the simulator, and the answer differs by dtype: **fp32 follows IEEE (496/496 NaN), bf16 does not (0/496)**. That split is the reason the claim is scoped to fp32.
- What is left to confirm on silicon is only whether the simulator's fp32 multiply matches the device's. If the device flushes there too, this does not reproduce and the issue can be closed — that single check settles it.

### Is this a regression?

No

### Regression Details

_No response_

### Logs & Diagnostics

No logs — there is no crash. The op returns NaN cleanly.

Same structural family, if it helps to batch the review:
- #51861 — `ttnn.hypot` returns `inf` because it computes `sqrt(a**2 + b**2)` literally. A composed op whose intermediate leaves the representable range where the final result would not.

### Priority

P2

### Impact

Affects `ttnn.tril` and `ttnn.triu` for any input containing `±inf` or `NaN` in the region being masked. The practical exposure is causal attention masking, which is where these ops are most used and where `-inf` padding is the convention — so the failing case sits close to the intended use rather than at an edge.

Because the output is `NaN` rather than a slightly wrong number, it propagates: one masked element poisons the softmax row that consumes it.

The fix is a re-composition using `ttnn::where`, which already exists, rather than a new kernel.

---

Written with AI assistance (Claude Opus 5 via Claude Code). I read the source, verified the IEEE behaviour and the mask counts myself, and I am responsible for the numbers and the claim as submitted.

### PR DESCRIPTION
**Pull request by ttino1312**
### Summary
Fixes #52038

`ttnn.tril` / `ttnn.triu` are composed as `multiply(input, mask)` where the mask
is a 0/1 index tensor. In IEEE-754, `0 * +/-inf = NaN` and `0 * NaN = NaN`, so any
infinity or NaN in the masked-out region came out as **NaN** instead of **0**
(fp32 path — bf16 happened to survive because of the narrower exponent).

Reproduced on ttsim: a 32x32 fp32 tensor of `-inf` turned 496/496 masked cells
into NaN under `tril`.

### Root cause
`ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp` — `tril`
(L308-318) and `triu` (L321-331) return `ttnn::multiply(input_a, index_l/u)`.

### Fix
Select instead of multiply — the 1/0 mask works as a `where` predicate, and the
masked region gets a true zero:

```cpp
return ttnn::where(index_l, input_a, 0.0f, output_mem_config.value_or(input_a.memory_config()));
```

Same change for `triu`. Behavior for normal inputs is unchanged (1xvalue in the
selected region, 0 outside).

### Validation
- New regression test `tests/ttnn/unit_tests/operations/eltwise/test_tril_triu_inf.py`:
  - full-`-inf` input: output must equal torch golden exactly, contain no NaN,
    and keep `-inf` in the selected region
  - normal-range input unchanged vs torch
  - both ops (`tril`, `triu`) x both dtypes (`bfloat16`, `float32`)

### CODE REVIEW

**Comment by mateusznowakTT** [comment]
Closing as there is earlier https://github.com/tenstorrent/tt-metal/pull/52145 already covering this.
